keps/sig-apps/4563-eviction-request-api/README.md

```diff
@@ -0,0 +1,1948 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are many issues with today's Eviction API and PodDisruptionBudgets that are of great concern
+to cluster administrators and application owners. These issues range from insufficient data safety,
+application availability, autoscaling to node draining issues.
+
+This KEP proposes to add a new declarative EvictionRequest API to manage the eviction of pods. Its
+mission is to allow for a cooperative termination of a pod, usually in order to run the pod on
+another node. If the owner of the pod does not cooperate, the eviction request will try to resort
+to a pod eviction (API-initiated Eviction).
+
+This new API can be used to implement additional capabilities around node draining, pod
+descheduling, autoscaling, or as a general interface between applications and/or controllers. It
+would provide additional safety and observability guarantees and prevent bad practices as opposed
+to just using the current Eviction API endpoint and PodDisruptionBudgets.
+
+## Motivation
+
+Many of today's solutions rely on API-initiated eviction as the goto-safe way to remove
+a pod from a node (kubectl drain, descheduler, cluster autoscaler, partially scheduler preemption).
+Unfortunately, this is done in an application agnostic way and can cause many problems.
+
+From an application owner or developer perspective, the only standard tool they have to protect
+them against eviction is a PodDisruptionBudget. This is sufficient in a basic scenario with a simple
+multi-replica application. The edge case applications, where this does not work are very important
+to the cluster admin or controllers managing workload distribution on nodes, as they can for
+example block the node drain. And, in turn, very important to the application owner, as the admin
+can then override the pod disruption budget and disrupt their sensitive application anyway.
+
+The major issues are:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
+   experiencing application downtime during the node drain. They cannot use PDBs with
+   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
+   needs high availability either, due to a preference for a simpler deployment model, lack of
+   application support for HA, or to minimize compute costs. Also, any automated solution needs
+   to edit the PDB to account for the additional pod that needs to be spun to move the workload
+   from one node to another. This has been discussed in issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
+   and in issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
+2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
```

Comment:
HPAs are fun here though because they can heal in ways are otherwise left to either deployment/statefulset controllers. So maybe it should be positioned as a consumer of this api?

Reply:
Yes, it is similar that it covers scaling of a single replica application, but the impact and followups are different. HPA covers 1-n replicas by default. So I think it is okay to keep it separate.

>  So maybe it should be positioned as a consumer of this api?

HPA is interesting because it can be both a requester (e.g. during scaling down) and an interceptor (e.g. during node drain). As is shown in the examples in the KEP.